Generate JWT realm from static config settings

justincr-elastic opened the pull request:
Refactor JWT util methods for reuse when creating JWT realm from static or random settings.

Small improvement to diagnostic logs during runtime authenticate. Filtering of JWKs and algorithms was not clear when trying to generate a static realm, and use it to perform a successful authentication.

For OIDC HMAC keygen, choose between random letters in the key, or random bytes wrapped by Base64. Always doing keygen via Base64(bytes) incorrectly implies Base64 is required. OIDC HMAC key encoding is specified as UTF8 in the specification, it is not limited to Base64.

justincr-elastic committed:
Generate static realm

Cleanup diagnostic logs to help generate, load, and get my static realm working.

Change OIDC HMAC keygen to include random letter keygen. Using Bas64(randomBytes) implies Base64 is required, but it is not for OIDC keys.

justincr-elastic committed:
Update docs/changelog/84323.yaml

elasticsearchmachine commented:
Hi @justincr-elastic, I've created a changelog YAML for you.

elasticmachine commented:
Pinging @elastic/es-security (Team:Security)

ywangd commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/jwt/JwtValidateUtil.java:
```diff
@@ -280,22 +283,41 @@ static void validateExpiredTime(final Date exp, final Date now, final long allow
+        assert jwks != null && jwks.isEmpty() == false : "Caller must provide a non-empty JWK list";
         final String id = jwt.getHeader().getKeyID();
         final JWSAlgorithm alg = jwt.getHeader().getAlgorithm();
         LOGGER.trace("JWKs [" + jwks.size() + "], JWT KID [ + id + ], and JWT Algorithm [" + alg.getName() + "] before filters.");
 
+        // If JWT has optional kid header, and realm JWKs have optional kid attribute, any mismatches JWT.kid vs JWK.kid can be ignored.
+        // Keep any JWKs if JWK optional kid attribute is missing. Keep all JWKs if JWT optional kid header is missing.
         final List<JWK> jwksKid = jwks.stream().filter(j -> ((id == null) || (j.getKeyID() == null) || (id.equals(j.getKeyID())))).toList();
-        LOGGER.trace("JWKs [" + jwksKid.size() + "] after KID [" + id + "||null] filter.");
+        LOGGER.debug("JWKs [" + jwksKid.size() + "] after KID [" + id + "||null] filter.");
```
I think these logs reallly should be trace level. They are step by step information. It does not fit [our guide](https://github.com/elastic/elasticsearch/blob/master/CONTRIBUTING.md#logging) for debug logging. It's also to verbose since it is for every request that autenticates with JWT.

justincr-elastic replied:
Do we have good guidelines for log levels? I have received different opinions on log levels. I keep having to toggle log levels, because I can't find any good guidelines. There seems to be a lot of unwritten rules and assumptions people make, and I don't know what they are.

Here is the issue. During authenticate, if we filtered JWKs in the realm, I want the customer to diagnose that type of authc failure themselves. In other words, I want to avoid support issues coming in for diagnosing this type of authc error.

There is a tradeoff for picking DEBUG vs TRACE, but I don't know what that tradeoff is.

tvernum replied:
https://github.com/elastic/elasticsearch/blob/master/CONTRIBUTING.md#log-levels

I agree this is a trace log.

justincr-elastic replied:
Resolved.

ywangd commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/jwt/JwtValidateUtil.java:
```diff
@@ -280,22 +283,41 @@ static void validateExpiredTime(final Date exp, final Date now, final long allow
-        for (final JWK jwk : jwksStrength) {
-            if (jwt.verify(JwtValidateUtil.createJwsVerifier(jwk))) {
-                return; // VERIFY SUCCEEDED
+        if (jwksStrength.isEmpty()) {
+            LOGGER.warn("JWKs [" + jwks.size() + "] usable [" + jwksStrength.size() + "]. See debug logs.");
+        } else {
```
If the list is empty, does that mean authentication will sure fail?
Also, I'd drop "See debug logs" which assumes debug logs are readily available but that is unlikely true.

justincr-elastic replied:
Any of the filters could reduce the JWK list to zero. The strength filter is the last filter. If none left at that point, there is no reason to go through the for-loop.

Example:
- Realm has HMAC 256-bit key.
- Current request is a JWT signed with HS512 algorithm.
- Trying the HMAC 256-bit key is pointless, it is too short for HS512 algorithm.
- In other words, only HMAC keys of 512-bit or larger need to be tried for JWTs with HS512 algorithm. If all HMAC keys are smaller, the validate for-loop is empty, and we should log a warning to tell customers that all HS512 JWTs have no chance of ever being authenticated by the realm.

justincr-elastic replied:
Changed from warn to trace. Resolved.

ywangd commented on x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/jwt/JwtRealmTests.java:
```diff
@@ -108,6 +104,122 @@ public void shutdown() throws Exception {
         terminate(this.threadPool);
     }
 
+    /**
+     * Generate a static JWT realm config, and request headers, which the realm can successfully authenticate.
+     *
+     * Print these artifacts, so they can be imported into this integration test project for verifying loading of JWT realm config:
+     *  - x-pack/plugin/security/qa/smoke-test-all-realms/build.gradle
+     *
+     * If something needs tweaking, change the code here and re-run it. The new artifacts can be used to update that smoke test project.
+     * @throws Exception Error if something goes wrong.
+     */
+    public void testCreateJwtSmokeTestRealm() throws Exception {
+        // static HMAC key (password UTF-8 bytes, for OIDC HMAC Key encoding compatibility)
+        final byte[] hmacKeyOidcBytes = "hmac-oidc-key-string-for-hs256-algorithm".getBytes(StandardCharsets.UTF_8);
+        final OctetSequenceKey hmacJwkOidc = new OctetSequenceKey.Builder(hmacKeyOidcBytes).build();
+
```
This method is not a test method. I think we should move it into its own helper file so it does not run with test suite but can be executed manually if needed to.

justincr-elastic replied:
I did think about moving it to another class, but it has dependencies on helper methods in JwtRealmTests, so I left it in there.

Also, it is actually doing authentication and authorization tests. The order of steps are:
1. Create the issuer, realm, user, and JWT
2. Run the authc/authz authentication test helper to verify everything works together
3. Print all of the artifacts

Printing the realm settings and request header are only possible if it passes the authc/authz tests, so it is actually a test.

justincr-elastic replied:
Do you still think it should be moved to a different test class? If yes, I will need to re-run it and put the changed output in the other PR.

justincr-elastic replied:
Created new JwtRealmTestCase, and split public test methods between two classes. Resolved.

ywangd commented on x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/jwt/JwtTestCase.java:
```diff
@@ -238,6 +239,55 @@ protected RealmConfig buildRealmConfig(
         };
     }
 
+    public static String printRealmSettings(final RealmConfig config, final boolean includeSecureSettings) {
+        final StringBuilder sb = new StringBuilder("\n===\nelasticsearch.yml settings\n===\n");
+        int numRegularSettings = 0;
+        for (final Setting.AffixSetting<?> setting : JwtRealmSettings.getNonSecureSettings()) {
+            final String key = RealmSettings.getFullSettingKey(config, setting);
+            if (key.startsWith("xpack.") && config.hasSetting(setting)) {
+                sb.append(key).append(": ").append(config.getSetting(setting)).append('\n');
+                numRegularSettings++;
+            }
+        }
+        if (numRegularSettings == 0) {
+            sb.append("Not found.\n");
+        }
```
Same for this method.

justincr-elastic replied:
Moved to JwtRealmGenerateTests. Resolved.

ywangd reviewed (commented):
I skimmed through the changes and I plan to look more closely later. But here are a few comments for a starter. Thanks!

justincr-elastic commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/jwt/JwtRealmSettings.java:
```diff
@@ -91,7 +98,7 @@ private JwtRealmSettings() {}
      * @return All non-secure settings.
      */
     public static Set<Setting.AffixSetting<?>> getNonSecureSettings() {
```
I think I made getNonSecureSettings() and getSecureSettings() public because they were called directly by test code in a previous PR. The test code is gone. I can change them to private.

justincr-elastic replied:
I can't change getNonSecureSettings() and getSecureSettings() to private. Best I can do it is make them package private.

Those methods are calls when printing realm settings. I loop through the first to print settings destined for elasticsearch.yml, and the second to print settings for elasticsearch-keystore. The contents within each block are printed in deterministic order, because those two settings methods internally use LinkedHashMap.

justincr-elastic replied:
Changed to package private. Resolved.

justincr-elastic commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/jwt/JwtRealmSettings.java:
```diff
@@ -131,8 +138,8 @@ private JwtRealmSettings() {}
      */
-    public static List<Setting.AffixSetting<SecureString>> getSecureSettings() {
-        return List.of(HMAC_JWKSET, HMAC_KEY, CLIENT_AUTHENTICATION_SHARED_SECRET);
+    public static Set<Setting.AffixSetting<SecureString>> getSecureSettings() {
```
I think I made getNonSecureSettings() and getSecureSettings() public because they were called directly by test code in a previous PR. The test code is gone. I can change them to private.

justincr-elastic replied:
Changed to package private. Resolved.

justincr-elastic commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/jwt/JwtRealmSettings.java:
```diff
@@ -272,4 +279,15 @@ private static void verifyNonNullNotEmpty(final String key, final List<String> v
         }
     }
+
+    public static String toYaml(final Settings settings) {
+        try (XContentBuilder builder = XContentBuilder.builder(XContentType.YAML.xContent())) {
+            builder.startObject();
+            settings.toXContent(builder, new ToXContent.MapParams(Collections.singletonMap("flat_settings", "true")));
+            builder.endObject();
+            return Strings.toString(builder);
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
```
I can remove it. I expected it to print regular and secure settings, but there is some obfuscation going on where it doesn't include the secure settings. I switched to alternate approach of looping through getNonSecureSettings() and getSecureSettings(), making their internal Set order deterministic, and using RealmConfig to retrieve the values. It was not very intuitive to have to do it that way, but it worked.

That reminds me, I can't change getNonSecureSettings() and getSecureSettings() to private. They are used in utility code. Best I can do it make them package private.

justincr-elastic replied:
Removed.

justincr-elastic commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/jwt/JwkValidateUtil.java:
```diff
@@ -44,18 +45,22 @@
+        LOGGER.debug("Algorithms [" + String.join(",", algsFiltered) + "] after remaining JWKs [" + jwksFiltered.size() + "] filter.");
 
+        if ((jwksFiltered.size() < jwks.size()) || (algsFiltered.size() < algs.size())) {
+            LOGGER.warn("JWKs [" + jwks.size() + "] usable [" + jwksFiltered.size() + "]. See debug logs.");
+            LOGGER.warn("Algorithms [" + String.join(",", algs) + "] usable [" + String.join(",", algsFiltered) + "]. See debug logs.");
```
There are two stages for filtering JWKs vs algs.
- Start up: Takes a list of algs from realm config (ex: HS256, RS256, ES256, ES512)
- Authenticate: Takes a single alg from JWT header (ex: ES256), allows ignoring realm algs for current request (i.e. ES512 has wrong curve to verify an ES256 signature)

I debated checking for size==0, or reduction in size. I went with reduction check. I tried to debug an issue in static realm config, and I realized if I had this much trouble, customers will end up creating a lot of support issues. That warning is all about avoiding SDHs.

Here is the issue. I created a static realm with an HMAC key in UTF8 string format, as per OIDC spec. I picked 20 chars (160-bits), but is too short for HS256. The key is filtered during startup. If PKC has at least one key, the realm will start. If I submit a JWT signed by that key with HS256, I may have that JWK and HS256 in the realm, and the realm started OK, but the request will never authenticate. Customers will blame the problem on a bug.

To avoid support issues coming in for that situation, realm startup needs to have a clear log if any JWKs as ignored due to filtering. If the customer searches for ERROR or WARN, they won't find any log message about problems loading that JWK. Using log level WARN will make it stand out.

tvernum replied:
That line of reasoning assumes that customers got through same process you did. I don't think that's necessarily true.

Take the case of PKC. The keys come from a JWKSet read from a file. The algorithms come from a realm setting. 
The JWKSet may contain multiple keys. The admin may intentionally configure the realm to support a subset of the algorithms for which those keys are usable. There's nothing wrong with:
```
xpack.security.authc.realms:
  jwt.my_jwt.pkc_jwkset_path: "/path/to/jwkset-with-many-rsa-keys.json"
  jwt.my_jwt.allowed_signature_algorithms: "RSA512"  
  # ^^ Reject JWTs signed with less than 512 bit keys
```

By logging a warning you are telling the admin that what they're done is wrong. But it's not. It _might_ be a misconfiguration, it's probably useful to tell them that some keys from the JWKTSet aren't usable, but it's not something to yell at them about.
I agree with `info`.

> realm startup needs to have a clear log if any JWKs as ignored due to filtering

I agree, but I don't think this log message achieves that. Anything that tells admins they need to enable debug logs in order to understand what's going on, does meet the "clear log" test for me.

justincr-elastic replied:
Changed two log messages to info. Resolved.

justincr-elastic commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/jwt/JwtRealm.java:
```diff
@@ -61,8 +61,8 @@ record JwksAlgs(List<JWK> jwks, List<String> algs) {}
     final CloseableHttpAsyncClient httpClient;
-    final JwtRealm.JwksAlgs jwksAlgsHmac;
-    JwtRealm.JwksAlgs jwksAlgsPkc; // reloadable
+    JwtRealm.JwksAlgs jwksAlgsHmac;
```
Reloadable is in a different milestone. Initially only PKC JWKSet was going to be reloadable, but I asked Tim and I think we agreed to make HMAC keys reloadable in future too. He said we support reloading for a few ES keystore settings, but we need to do more of that, so I think it makes sense to remove the final qualifier.

tvernum replied:
If it's reloadable in the future we can drop `final` then. There's no value in doing that now (we don't need to guess on what we will implement in the future & how it would be implemented)

justincr-elastic replied:
I will make it final. It means I have to change the signature of two methods to return those records so I can set final in the constructor.

justincr-elastic replied:
Resolved.

tvernum commented on docs/changelog/84323.yaml:
```diff
@@ -0,0 +1,5 @@
+pr: 84323
+summary: Small cleanup to help generate JWT realm from static config settings
+area: Authentication
+type: feature
```
This PR should be labelled `>non-issue` rather than `>feature`.

This PR doesn't introduce a brand new feature.
Since the changes it makes are too an unreleased feature, the PR itself is `>non-issue` because people upgrading from an earlier version of ES to the version that contains this PR don't need to know about this specific PR, just the overall PR that introduced JWT.

justincr-elastic replied:
Changed to >non-issue. Resolved.

justincr-elastic committed:
Package private for getNonSecureSettings() and getSecureSettings()

justincr-elastic committed:
Private getNonSecureSettings() and getSecureSettings(). Move printRealmSettings to JwtRealmsTest from JwtTestCase.

justincr-elastic commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/jwt/JwtRealmSettings.java:
```diff
@@ -80,7 +87,7 @@ private JwtRealmSettings() {}
      * @return All secure and non-secure settings.
      */
     public static Set<Setting.AffixSetting<?>> getSettings() {
-        final Set<Setting.AffixSetting<?>> set = Sets.newHashSet();
+        final Set<Setting.AffixSetting<?>> set = new LinkedHashSet<>();
```
Each realm uses Set collection for settings. There is no restriction to HashSet vs LinkedHashSet?

justincr-elastic replied:
Changed all LinkedHashSet to HashSet.

justincr-elastic committed:
Changed two log messages to info.

justincr-elastic committed:
Change validateSignature logs from debug to trace. Remove "See debug logs."

justincr-elastic committed:
Fix code analysis warning about size always 0 in log message.

justincr-elastic committed:
Add final qualifier to JwtRealm records jwksAlgsHmac and jwksAlgsPkc.

The method signature of two methods was required to change to accommodate the final qualifier change.

justincr-elastic committed:
Delete docs/changelog/84323.yaml

justincr-elastic committed:
Changed isMatch() logs from warn to debug.

justincr-elastic committed:
Merge remote-tracking branch 'fork/feature/jwt-realm-generate-static-realm' into feature/jwt-realm-generate-static-realm

justincr-elastic commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/jwt/JwkValidateUtil.java:
```diff
@@ -70,11 +75,29 @@
+                final boolean isMatch = bits >= min;
+                if (isMatch == false) {
+                    LOGGER.warn("RSA JWK [" + bits + "] bits too small for algorithm [" + algorithm + "] minimum [" + min + "].");
+                }
```
Resolved.

justincr-elastic commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/jwt/JwkValidateUtil.java:
```diff
@@ -70,11 +75,29 @@
+                final boolean isMatch = allowed.contains(curve);
+                if (isMatch == false) {
+                    LOGGER.warn("EC JWK [" + curve + "] curve not allowed for algorithm [" + algorithm + "] allowed " + allowed + ".");
+                }
```
Resolved.

justincr-elastic committed:
Remove JwtRealmSettings.toYaml()

justincr-elastic commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/jwt/JwkValidateUtil.java:
```diff
@@ -70,11 +75,29 @@
+                final boolean isMatch = bits >= min;
+                if (isMatch == false) {
+                    LOGGER.warn("HMAC JWK [" + bits + "] bits too small for algorithm [" + algorithm + "] minimum [" + min + "].");
+                }
```
Changed from warn to debug. Resolved.

justincr-elastic committed:
Changed warn to trace in validateSignature()

justincr-elastic committed:
Rename JwtRealmTests to JwtRealmTestCase. Move test methods to JwtRealmAuthenticateTests or JwtRealmGenerateTests. The generate smoke test realm code is now in its own class.

justincr-elastic committed:
Changed all LinkedHashSet to HashMap.

justincr-elastic committed:
Revert "Delete docs/changelog/84323.yaml"

This reverts commit 0ad680592ba6f3b0c0bf27c9b124aebd8582c475.

justincr-elastic committed:
Change 84323.yml from feature to non-issue.

justincr-elastic committed:
Delete docs/changelog/84323.yaml

justincr-elastic commented:
All requested changes done. Ready for review again.

ywangd commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/jwt/JwtValidateUtil.java:
```diff
@@ -280,22 +283,41 @@ static void validateExpiredTime(final Date exp, final Date now, final long allow
-        for (final JWK jwk : jwksStrength) {
-            if (jwt.verify(JwtValidateUtil.createJwsVerifier(jwk))) {
-                return; // VERIFY SUCCEEDED
+        if (jwksStrength.isEmpty()) {
+            LOGGER.trace("JWKs [" + jwks.size() + "] usable [0].");
+        } else {
```
If `jwksStrength` is empty, the method throws an Exception which contains the information about size of `0`. So I think this log is redundant. 
However, I think it is valuable to have a debug log that tells the final state after all filtering similar to `filterJwksAndAlgorithms`, e.g. `JWKs usable [3] out of [5]`.

justincr-elastic replied:
Are you saying remove log at 315? And change log at line 312 to debug? I thought you wanted logs in this method to be trace.

https://github.com/elastic/elasticsearch/blob/bd85f0cff227bb87589c66b7fa5a9b28cfc91505/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/jwt/JwtValidateUtil.java#L312

justincr-elastic replied:
Nits on logging in previous PR led to delays because of extra cycles. I added a task to revisit logs holistically in a later PR.

ywangd replied:
Having a separate task for revisiting logs is a good idea.

ywangd commented on x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/jwt/JwtRealmGenerateTests.java:
```diff
@@ -0,0 +1,207 @@
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.is;
+
+public class JwtRealmGenerateTests extends JwtRealmTestCase {
+    private static final Logger LOGGER = LogManager.getLogger(JwtRealmGenerateTests.class);
+
+    /**
+     * Generate a static JWT realm config, and request headers, which the realm can successfully authenticate.
+     *
+     * Print these artifacts, so they can be imported into this integration test project for verifying loading of JWT realm config:
+     *  - x-pack/plugin/security/qa/smoke-test-all-realms/build.gradle
+     *
+     * If something needs tweaking, change the code here and re-run it. The new artifacts can be used to update that smoke test project.
+     * @throws Exception Error if something goes wrong.
+     */
+    public void testCreateJwtSmokeTestRealm() throws Exception {
```
My previous suggestion is that this method is not a test (because it does not test anything) and hence is better to be a helper class/method instead of running as part of the test suite. I now notice how it utilize the super class methods overall the places. It's not easy to relocate them and probably not worth it. I am happy for this to stay as is.

ywangd reviewed (approved):
LGTM pending CI pass 

I left a few more comments but nothing major.

justincr-elastic commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/jwt/JwtRealm.java:
```diff
@@ -125,97 +127,78 @@ public JwtRealm(final RealmConfig realmConfig, final SSLService sslService, fina
+                jwksHmac = List.of(hmacKey);
+            }
+            // Filter JWK(s) vs signature algorithms. Only keep JWKs with a matching alg. Only keep algs with a matching JWK.
+            final List<String> algs = super.config.getSetting(JwtRealmSettings.ALLOWED_SIGNATURE_ALGORITHMS);
+            final List<String> algsHmac = algs.stream().filter(JwtRealmSettings.SUPPORTED_SIGNATURE_ALGORITHMS_HMAC::contains).toList();
```
Your understanding about supported vs allowed sounds right. Supported is all algorithms that ES supports, and allowed is a subset that the user will allow.

Originally, I did throw an exception if I found any JWKs or algorithms that were orphaned. Ioannis asked me to allow it, so I changed it to logs. Now, I only throw an exception if all JWKs and algorithms are filtered out.

Examples:
1. Configure PKC JWKSet with algorithm HS256. They are not usable together.
2. Configure HMAC key as 31-bytes with algorithm HS256. The key is too short for that algorithm.
3. Configure HMAC key as 32-bytes with algorithm HS-256. If KeyUse=ENCRYPTION, they are not usable together.

justincr-elastic replied:
Thank you for adding the extra context, Ioannis.

justincr-elastic committed:
Add JwksAlgs.isEmpty method.

justincr-elastic commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/jwt/JwtRealm.java:
```diff
@@ -125,97 +127,78 @@ public JwtRealm(final RealmConfig realmConfig, final SSLService sslService, fina
+    private void verifyAnyAvailableJwkAndAlgPair() {
+        assert this.jwksAlgsHmac != null : "HMAC not initialized";
+        assert this.jwksAlgsPkc != null : "PKC not initialized";
+        if (((this.jwksAlgsHmac.jwks.isEmpty()) && (this.jwksAlgsPkc.jwks.isEmpty()))
+            || ((this.jwksAlgsHmac.algs.isEmpty()) && (this.jwksAlgsPkc.algs.isEmpty()))) {
```
Added isEmpty method, but I cannot use it here because the above check is (HMAC.jwks && PKC.jwks) || (HMAC.algs && PKC.algs). I may find a different use for the isEmpty helper later.

justincr-elastic committed:
Removed log if no JWKs remaining. Changing last filter to debug.

justincr-elastic commented:
@elasticmachine update branch

elasticmachine commented:
merge conflict between base and head

justincr-elastic commented:
@elasticmachine update branch

justincr-elastic committed:
Merge branch 'feature/jwt-realm-generate-static-realm' of https://github.com/justincr-elastic/elasticsearch into feature/jwt-realm-generate-static-realm

justincr-elastic commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into feature/jwt-realm-generate-static-realm

justincr-elastic committed:
Improve log levels.

justincr-elastic commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into feature/jwt-realm-generate-static-realm

justincr-elastic commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into feature/jwt-realm-generate-static-realm

justincr-elastic commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into feature/jwt-realm-generate-static-realm

justincr-elastic commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into feature/jwt-realm-generate-static-realm
